provider/cluster_rosa_classic_resource.go

```diff
@@ -1326,47 +1340,55 @@ func (r *ClusterRosaClassicResource) upgradeClusterIfNeeded(ctx context.Context,
 	if err != nil {
 		return fmt.Errorf("failed to parse desired cluster version: %v", err)
 	}
-	if currentVersion.GreaterThanOrEqual(desiredVersion) {
+	if currentVersion.GreaterThan(desiredVersion) {
 		tflog.Debug(ctx, "No cluster version upgrade needed.")
+		if requestedVersionChanged {
+			// User changed the version they want, but actual is higher. We
+			// don't support downgrades.
+			return fmt.Errorf("cluster version is already above the requested version")
+		}
 		return nil
 	}
+	cancelingUpgradeOnly := desiredVersion.Equal(currentVersion)
 
-	// Make sure the desired version is available
-	versionId := fmt.Sprintf("openshift-v%s", state.CurrentVersion.Value)
-	if !state.ChannelGroup.Unknown && !state.ChannelGroup.Null && state.ChannelGroup.Value != ocm.DefaultChannelGroup {
-		versionId += "-" + state.ChannelGroup.Value
-	}
-	availableVersions, err := upgrade.GetAvailableUpgradeVersions(ctx, r.versionCollection, versionId)
-	if err != nil {
-		return fmt.Errorf("failed to get available upgrades: %v", err)
-	}
-	found := false
-	for _, v := range availableVersions {
-		sem, err := semver.NewVersion(v.RawID())
-		if err != nil {
-			return fmt.Errorf("failed to parse available upgrade version: %v", err)
+	if !cancelingUpgradeOnly {
```

Comment:
That function become to be complicated - Can you refactor that code and move it to another function?